runtime/gc_glue_java/FlattenedArrayObjectScanner.hpp

```diff
@@ -0,0 +1,250 @@
+	 * @param[in] scanPtr pointer to the array cell where scanning will start
+	 * @param[in] endPtr pointer to the array cell where scanning will stop
+	 * @param[in] scanMap The first scan map
+	 * @param[in] elementSize The size of every element
+	 * @param[in] elementClazzPtr Pointer to the element's class
+	 * @param[in] flags Scanning context flags
+	 */
+	MMINLINE GC_FlattenedArrayObjectScanner(
+		MM_EnvironmentBase *env
+		, omrobjectptr_t arrayPtr
+		, fomrobject_t *basePtr
+		, fomrobject_t *limitPtr
+		, fomrobject_t *scanPtr
+		, fomrobject_t *endPtr
+		, uintptr_t scanMap
+		, uintptr_t elementSize
+		, J9Class *elementClazzPtr
+		, uintptr_t flags)
+	: GC_HeadlessMixedObjectScanner(env, elementClazzPtr, scanPtr, flags | GC_ObjectScanner::indexableObject)
+	, _env(env)
+	, _elementClazzPtr(elementClazzPtr)
+	, _indexableScanner(env, arrayPtr, basePtr, limitPtr, scanPtr, endPtr, scanMap, elementSize, flags)
+	{
+		/* The HeadlessMixedObjectScanner will setNoMoreSlots(), 
+		 * causing us to miss other elements of the array */
+		setMoreSlots();
+		_typeId = __FUNCTION__;
+	}
+
+	MMINLINE void
+	initialize(MM_EnvironmentBase *env, J9Class *elementClazzPtr)
+	{
+		GC_HeadlessMixedObjectScanner::initialize(env, elementClazzPtr);
+
+		// TODO: this is protected but doesnt do anything
+		// _indexableScanner.initialize(env);
+
+		fprintf(stderr, "!!!  element (%p,%p)\n", _mapPtr, _endPtr);
+		fprintf(stderr, "!!!    mapPtr=%p, slotmap=%#lx, leafmap=%#lx\n", _mapPtr, _scanMap, _leafMap);
+	}
+
+public:
+
+	/**
+	 * @param[in] env The scanning thread environment
+	 * @param[in] objectPtr pointer to the array to be processed
+	 * @param[in] allocSpace pointer to space within which the scanner should be instantiated (in-place)
+	 * @param[in] flags Scanning context flags
+	 * @param[in] splitAmount If >0, the number of elements to include for this scanner instance; if 0, include all elements
+	 * @param[in] startIndex The index of the first element to scan
+	 */
+	MMINLINE static GC_FlattenedArrayObjectScanner *
+	newInstance(MM_EnvironmentBase *env, omrobjectptr_t objectPtr, void *allocSpace, uintptr_t flags, uintptr_t splitAmount, uintptr_t startIndex = 0)
+	{
+		GC_FlattenedArrayObjectScanner *objectScanner = (GC_FlattenedArrayObjectScanner *)allocSpace;
+		GC_ArrayObjectModel *arrayObjectModel = &(env->getExtensions()->indexableObjectModel);
+		J9Class *clazzPtr = J9GC_J9OBJECT_CLAZZ(objectPtr, env);
+		J9ArrayClass *j9ArrayClass = (J9ArrayClass *) clazzPtr;
+
+		// TODO are these always the same? 
+		Assert_MM_true(j9ArrayClass->componentType == j9ArrayClass->leafComponentType);
+
+		J9Class *elementClass = j9ArrayClass->componentType;
+
+		omrarrayptr_t arrayPtr = (omrarrayptr_t)objectPtr;
+		uintptr_t sizeInElements = arrayObjectModel->getSizeInElements(arrayPtr);
+		uintptr_t elementSize = J9ARRAYCLASS_GET_STRIDE(clazzPtr);
+		env->getExtensions()->mixedObjectModel.getSizeInBytesWithoutHeader(elementClass);
```

Comment:
Is it a left over? Was it plan to have assertion to compare element size from stride and from element class?

Reply:
Yeah, its a left over from before I was using stride. woops!